### ISSUE TITLE
Unserialize Drawings

### ISSUE THREAD

**Issue by workingbuddy10** [amcharts/amcharts5#1280]
I am using unserializeDrawings in one of my feature where I pass the data of the drawing tool and whenever the user clicks on the any tool it unserialize the Drawing of that particular/selected only.

And when user removes/un-selects the drawing tool the drawing should be removed.  I am getting stuck in this part only.

To remove the unserializeDrawings user can use eraser/delete all drawing option. What if I want to remove the drawings on the un-check of the tool option WITHOUT again un-serializing Drawings.

Image for the better reference how the feature looks like : 

![image](https://github.com/amcharts/amcharts5/assets/137867461/99bcf2dd-6196-42c1-ac06-c4c0bef9ad80)

![image](https://github.com/amcharts/amcharts5/assets/137867461/7a6730f8-6ed8-4805-bc3d-0e3833fb3ded)

And when I remove the Line option the drawing should be removed :

![image](https://github.com/amcharts/amcharts5/assets/137867461/6ebf13fa-ad44-4b46-9f86-65d24164500b)

**Comment by workingbuddy10** [comment]
@martynasma  Please any help in this would be appreciated.

**Comment by martynasma** [comment]
We'll introduce a public method on DrawingSeries to remove its drawings by index.

Will send more details when the new version is out.

**Comment by martynasma** [comment]
Hey, so 5.7.7 is out. It adds a new method on all drawing series: `disposeIndex(index)`.

Calling it will dispose ant N-th drawing item in the series.

You can find specific drawing series like this:

```JavaScript
var lineSeries;
mainPanel.drawings.each(function(series) {
  if (series.className == "SimpleLineSeries") {
    lineSeries = series;
  }
})
lineSeries.disposeIndex(1);
```

Hope it helps.

**Comment by workingbuddy10** [comment]
And what about unserializeIndicators, how will differentiate between unserializeIndicators and Indicators used by user?

It would be better if there is something in-built delete button for deleting the unserialize drawings/indicators. What do you suggest?
@martynasma

**Comment by martynasma** [comment]
There is an eraser button in Draw menu for erasing indicators. You can also enable erasing mode using API: https://www.amcharts.com/docs/v5/charts/stock/toolbar/drawing-control/#Toggling_eraser

Indicators can either be removed using their legend, or by directly disposing indicator instance.

All indicators are registered in [`StockChart.indicators`](https://www.amcharts.com/docs/v5/reference/stockchart/#indicators_property).

**Comment by workingbuddy10** [comment]
When I use Delete ALL Drawings button, it deletes the unserializeDrawings  drawings too!

I only to delete the unserializeDrawings  by method you told me of _disposeIndex_ .

I want : 
 - User made drawings should be deleted by the Earaser & Delete All drawings button ONLY.
 - And unserializeDrawings should be deleted by disposeIndex method which you introduced ONLY.

**Comment by workingbuddy10** [comment]
And the second issue I am facing in a case is : 

When I draw a drawing (Line) on the unserializeDrawing ( which is also a line) then dispose index is not deleting the drawing.

The Line which is marked in Yellow is unserializeDrawing and the Red marked line is user drawn.
![image](https://github.com/amcharts/amcharts5/assets/137867461/fc3dfe33-544a-40fb-9d3d-d5d8def6187b)

I am deleting like this :
```
     deleteAutoIndicator(deleteInd) {
        
        this.chartInstance.mainPanel.drawings.each(function (series) {
            if (series.className == "SimpleLineSeries") {
                deleteInd = series;
            }
        })
        deleteInd.disposeIndex(2);

    }
```

Can you please open the issue again it would be good to track it.
@martynasma

**Comment by workingbuddy10** [comment]
Can I delete the drawing by passing the toolname? 
Instead of passing the index can we delete it by passing the __tool name?

```
var lineSeries;
mainPanel.drawings.each(function(series) {
  if (series.className == "SimpleLineSeries") {
    lineSeries = series;
  }
})
lineSeries.disposeIndex(1);
```

**Comment by workingbuddy10** [comment]
> And the second issue I am facing in a case is :
> 
> When I draw a drawing (Line) on the unserializeDrawing ( which is also a line) then dispose index is not deleting the drawing.
> 
> The Line which is marked in Yellow is unserializeDrawing and the Red marked line is user drawn. ![image](https://private-user-images.githubusercontent.com/137867461/298088651-fc3dfe33-544a-40fb-9d3d-d5d8def6187b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5MTk4MDMsIm5iZiI6MTcwNTkxOTUwMywicGF0aCI6Ii8xMzc4Njc0NjEvMjk4MDg4NjUxLWZjM2RmZTMzLTU0NGEtNDBmYi05ZDNkLWQ1ZDhkZWY2MTg3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyMlQxMDMxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYjczYWMzMzk1NDdlMmM4M2RlZmVlNDBlZDU0YTYxNTI4ODkxMmEwMmU1MGM2ZDNhNjgyMjJiYmU5YWIwNmRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vFJH6kjfmHZWfrxKq6EElXdnz55raVa6w7VwZxw0ek8)
> 
> I am deleting like this :
> 
> ```
>      deleteAutoIndicator(deleteInd) {
>         
>         this.chartInstance.mainPanel.drawings.each(function (series) {
>             if (series.className == "SimpleLineSeries") {
>                 deleteInd = series;
>             }
>         })
>         deleteInd.disposeIndex(2);
> 
>     }
> ```
> 
> Can you please open the issue again it would be good to track it. @martynasma

Please let me know about this too
@martynasma

**Comment by martynasma** [comment]
Not, via panel.

`DrawingControl` has a list of drawing series in a private property. We'll expose it publicly in the next version so you can access it like `drawingControl.drawingSeries["Line"]`.

**Comment by workingbuddy10** [comment]
And How to restrict the Delete All button, from deleting the unserializeDrawing.

@martynasma

**Comment by workingbuddy10** [comment]
And how to distinguish which drawing is unserializeDrawing  & which one is user made.

As If I am able to achieve this then most of the problem will be solved, as I am uanble to target the user made drawings & unserializeDrawings.

**Comment by martynasma** [comment]
There's no way to distinguish them programmatically, I'm afraid.

**Comment by workingbuddy10** [comment]
> @martynasma And How to restrict the Delete All button, from deleting the unserializeDrawing?

Can we achieve this?

**Comment by workingbuddy10** [comment]
> There's no way to distinguish them programmatically, I'm afraid.

Do you think it should there in amcharts ?

As it will solve : 
- Deleting the unserializeDrawing particularly without hindering the user made drawings.

And unserializeDrawings is something that a user made in the past, stored it, and later retrieved. It can be totally different depending on which user made the drawings at the present.
So other features, like deletion and other properties, should not conflict with the current drawings, which are drawn by the user.

I am eager to know your opinion about the ambiguity.
@martynasma

**Comment by martynasma** [comment]
> Do you think it should there in amcharts ?

Possibly. However, at this time we do not have immediate plans to implement them.

**Comment by martynasma** [comment]
Implemented in [5.8.0](https://github.com/amcharts/amcharts5/releases/tag/5.8.0).

## [5.8.0] - 2024-02-01

### Added
- New container type `ZoomableContainer` added, which allows adding zoom capabilities to virtually any chart. [More info](https://www.amcharts.com/docs/v5/concepts/common-elements/containers/#Zoomable_container).
- New class `ZoomTools` added. Can be used to quickly add zoom support for elements compatible with `IZoomable` interface, e.g. `ZoomableContainer`.
- New read-only property of `DrawingControl`: `drawingSeries`. Contains an object where key is a drawing tool name and values are array with references to actual drawing series.

### Changed
- Default value for `strokeOpacity` for the grid line of a `StockChart` changed from `0` to `0.4` so that it would differ from regular grid lines.

### Fixed
- Resizing stock chart's panels after moving them up/down was not working properly.
- `Hierarchy.addChildData` was duplicating node's children which were added before.
- If `topDepth` was set to `1` on `Partition`, initially the chart was showing not all levels of nodes.
- When clicking on particular partitions of a `Partion` chart, zoom animation was not playing.
- `DateRangeSelector` was showing extra day in automatically-calculated end date.
- `DateRangeSelector` was zooming to the start of the selected day, instead of end.
- Tooltip background color passed to `HeatLegend.showValue(value, text, color)` was being ignored.
- `ClusteredPointSeries` was not showing charts if they were added to clustered bullets.
- Fixed issue with `DrawingControl` which was unnecessarily duplicating drawing series with each tool selection / API call.
- Using "1M" button in `PeriodSelector` was not always selecting full month.
- Restoring drawings with `DrawingControl.unserializeDrawings()` will now correctly update indexes of the restored drawings based on current drawing counter.
- In some cases `StockChart` was taking a data item which was just outside of zoom range when calculating aggregate values for its series, affecting percent change values and some other aggregates.
- Legend value was not being updated when series axis' range changed. This caused `legendRangeLabelText` to not being updated (if set) when chart was being zoomed-in/out.
- If a type of `StockChart` series was changed with data grouping enabled, some duplicate candlesticks or ohlc sticks were displayed until first zoom.
- In some rare cases chart would show incorrect last segment of a `LineSeries` when in percent mode.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
